### ISSUE TITLE
Warn or abort download if the downloading image would bring device in out-of-disk-space state

### ISSUE THREAD

**Issue by pcarranzav** [resin-io/resin-supervisor#29]

**Comment by pdcastro** [comment]
It has been suggested in [this forum comment](https://forums.balena.io/t/looping-download-when-no-space-left-on-device-error/6771/9) that deciding whether there is enough disk space is a difficult or expensive operation because delta downloads may mean that a "2GB image" actually only needs 2KB extra disk space.

But what the user complains about in the forum is subtly but importantly different, and would be a cheap operation. Go ahead with attempting the update, without trying to determine whether there is enough disk space. Then if it fails with an error related to disk space, don't try again until or unless the circumstances have changed. The user's point is that the operation is retried again and again and again... All the while logging that there isn't enough disk space. The heuristic check might be:

- Take note of how much free disk space there was before the update attempt -- a "naive and cheap" measure like what's reported by `df`, without taking deltas into account.
- Try the update. If it fails because of disk space, then only retry the operation if:
  - The naive/cheap disk space measure is now meaningfully larger than the measure when the previous attempt failed, or
  - The target state (downloading image) has changed.

**Comment by pcarranzav** [comment]
Interesting idea, and I think it could work if we add the concept of free space to the supervisor's internal representation of the device state. It is indeed easier than preemptively aborting, but it's still not a simple, easy change though, as it requires detecting that the error was actually disk space (which doesn't always fail in the same way), measuring disk space when this happens, and then changing the logic in ApplicationManager to return "noop" actions when in this state.

**Comment by pdcastro** [comment]
> _it requires detecting that the error was actually disk space_

Indeed. The user reports that _"it generates an ‘error no space left on device’"_ (repeatedly), so at some level in the error / logging stack, this information is surfacing.

**Comment by pcarranzav** [comment]
Yeah, looking for "no space left on device" is one failure mode we can easily catch, but what I mean is that this is not always the case - it's not unusual that the device will be out of space and the actual error thrown will be something else (e.g. balena-engine will be unable to create an intermediate file during download, and complain about that, without explicitly mentioning disk space). What the user saw was the "lucky" case when the error is explicit about the space issue.

That being said, we can definitely start with the easy case and later add more failure modes / log messages that we catch.

There's also been alternative suggestions though, like for *any* error (rather than, say, a connectivity hiccup), put an upper limit on the number of times the supervisor will retry an update. After this, the supervisor would enter a "failed" state that needs to be manually reset. This could work as a catch-all to avoid crazy bandwidth consumption cases, with the obvious downside that requiring manual intervention can be far from ideal, especially on big fleets.

Probably worth an architecture call discussion :)

**Comment by CameronDiver** [comment]
@pcarranzav I really like you're idea on the failure mode. We could even mirror this in the dashboard, so a user knows that their device is misbehaving. I'm going to raise it for an arch call.

**Comment by roman-mazur** [comment]
Related: #1033
